### PR TITLE
Add Grid, Box, Figure components

### PR DESCRIPTION
**Pull request by taptapdan**
## Background

Last week, there was a display bug on mobile reported by Alejandro. I put in a [quick fix](https://github.com/PERTS/triton/pull/1751) but this ended up breaking most other uses of Multicolumn and Figure components. In order to give me more time to come up with a more universal solution, I [reverted the change](https://github.com/PERTS/triton/pull/1754).

The revert re-broke the original mobile display bug, but it was one instance vs most.

### Grid/Box

In order to come up with a more universal layout solution, I've added Grid and Box components.

```jsx
<Grid>
  <Box cols={6}>Two equal width columns</Box>
  <Box cols={6}>Two equal width columns</Box>
</Grid>

<Grid>
  <Box cols={3}>Different sized columns</Box>
  <Box cols={6}>Different sized columns</Box>
  <Box cols={3}>Different sized columns</Box>
</Grid>

<Grid>
  <Box cols={6} colsMd={12}>Responsive columns</Box>
  <Box cols={6} colsMd={12}>Responsive columns</Box>
</Grid>
```

### Figure

The new Figure component defaults to being responsive, expanding to 100% width of its parent container. You can set `responsive={false}` if you'd like to disable that.

This component is a little less thought out. I'm not 100% sure on its use, I'm not completely following its use over in triton. For example, there are cases where it seems like we could just use `` tags. So I tried to be as light handed as possible with this component, just doing the minimum need to get it working with Grid/Box.

### LoremText and FakeText

For use in Storybook stories.

## Testing

Added Storybook stories on usage. See Grid and Figure stories.

A number of the new stories have responsive examples, so you can either resize your browser or use the Storybook built-in preview window size selector.

Screen Shot 2020-07-22 at 2 18 50 PM

### CODE REVIEW

File: lib/components/Card/index.js [hunk 1]
```diff
@@ -72,6 +47,49 @@ CardContentStyled.defaultProps = {
   theme,
 };
 
+const CardStyled = styled.div`
+  // TODO Decide how to handle placing Card components within Grid/Box layouts.
+  // We'd like Cards to be able to stand on their own, but also be compatible
+  // within the Grid system. This margin-bottom rule, when placed within Grid
+  // results in extra spacing. Removing this rule will break current usage of
+  // this component.
```
Comment (chris-perts) [inline lib/components/Card/index.js:55]: proposal: override the margin-bottom rule within `Box`:

CardStyled:

`margin-bottom: ${(props) => props.theme.units.fieldPadding};`

Box:

```
${Card} {
  margin-bottom: 0;
}
```

**Comment by chris-perts** [comment]
> I'm not completely following its use over in triton. For example, there are cases where it seems like we could just use  tags.

It's just semantic. Any styles associated with it are unimportant, other than it's a block element. I think I added some sizings that were convenient because at the time I didn't want to build out a full grid system.

> Graphics that are just graphics should still be listed as . The reason regular graphics deserve different treatment is because they're not necessary to the content on the page... Instead of nesting an image and a caption you've put in a <p> within a <div> with a class of "image" or something similar, you can group all image like things together inside of a <figure>...Use <figure> if the content is essential, but it's exact position on the page is important.

https://www.pluralsight.com/blog/creative-professional/whats-difference-image-figure-tags-html5

**Comment by taptapdan** [comment]
Leaving some incomplete notes here on Grid system options for our conversation later...

## Existing Options

- My preference is that we use a Grid system that has an existing React component library (not just stylesheets) so we don't need to create our own. Needing to do so would seem like a time/maintenance loss.
- My preference is that we use a Grid system that is intelligent about importing of stylesheets. Needing to import styles for 30+ components (for example) to utilize 1 Grid component would seem like a user experience loss.
- My preference is that we use a Grid system that is compatible with Create React App.
- [Skeleton Grid](https://github.com/dhg/Skeleton) appears to be just a stylesheet. A quick search did not turn up a React component library.
- [Simple Grid](http://simplegrid.io/) appears to be just a stylesheet. A quick search did not turn up a React component library.
- [Material UI Grid](https://material-ui.com/components/grid/)'s custom styling system is awful and not compatible with how we're currently maintaining our themes. If there's no better solution, perhaps we could create transformation functions that would allow us to utilize the library.
- [Semantic UI](https://react.semantic-ui.com/) I like the look and simplicity of this library the best, but it's either eject CRA, or import all the stylesheets and overwrite them with custom css styles (`.container .grid .row { our-style-override: here }`). I don't love the idea of maintaining some component style work in css files and some in styled-components.
- Ant Design has a [Grid system](https://ant.design/components/grid). Without ejecting CRA, it requires importing the entire library stylesheet, but at least their [theming](https://ant.design/docs/react/use-with-create-react-app) seems to allow a slightly easier method of overriding the base theme. It isn't accessible from within the app, it's a webpack config phase override, but the config is javascript so I'd imagine we could probably access our components/theme. Unfortunately, based on their examples and 60 seconds of playing, it didn't appear that their responsive grid actually drops boxes down to the next line at breakpoints. It seems to only shrink them down. Not sure if this is something I'm not understanding or a bug or it's just not what we want.

## Naming

We can definitely rename it from Grid to something else. I avoided "Flex" because there was already a Flex component (that the Grid system would deprecate, but I didn't want to take the time to do that at the moment) and went with Grid because that's what the internet calls it (see the above 5 examples of popular grid system solutions). But just let me know what you want and we'll "make it so."

**Comment by chris-perts** [comment]
👍 for the research, that's very useful. I'm fully satisfied that there's no drop-in react-based solution (and I've tried `react-bootstrap` and while it _is_ a component library I remember being frustrated with its lack of support, old issues and such).

Now I'm trying to problem-solve how bullet-proof your CSS. Can we compare your width rules to the kinds of rules Bootstrap uses for the same thing? In other words, maybe we can save ourselves time by imitating third-party stylesheets in a minimal way.

Poking around their code, they're doing this:

```
.col-2 {
  -ms-flex: 0 0 16.666667%;
  flex: 0 0 16.666667%;
  max-width: 16.666667%;
}

.col-3 {
  -ms-flex: 0 0 25%;
  flex: 0 0 25%;
  max-width: 25%;
}

/* et cetera */
```

That's _similar_ to what you're doing, but not exactly the same:

* width vs. max-width
* using `auto` vs. pixels in `flex`

You're stronger in CSS than I am; do you recognize any advantages here? Costs to changing it?

----

You're right that people seem to play fast and loose with the term "css grid". How about `Row` and `Col`, to imitate Bootstrap as closely as possible, especially since the props you've specified use "col" in them, also like bootstrap. Or maybe `GridRow`? I basically want to leave room in the namespace for a component that actually is a Grid, a la `display: grid`.

**Comment by taptapdan** [comment]
Yup. `Row` and `Col` work.

**Comment by chris-perts** [comment]
Note: @taptapdan originally pulled his column styles (flex and max-width and such) from [Quasar](https://quasar.dev/layout/grid/row).

**Comment by taptapdan** [comment]
- fbadb6b renamed Grid to Row and Box to Col
- 5ca9bb8 added comment on row/col quasar source

Updating so we can get this work into Triton so I can utilize the changes for the ascend20 modules.

**Pull request merged**